flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/BatchJobRecoveryTest.java

```diff
@@ -257,6 +260,11 @@ void testRecoverFromJMFailover() throws Exception {
                 createScheduler(deserializeJobGraph(serializedJobGraph));
         startSchedulingAndWaitRecoverFinish(newScheduler);
 
+        waitUntilExecutionVertexState(
```

Comment:
Why is it needed?

IIUC, the status of source vertices should have recovered before the recovery process finishes.